### PR TITLE
UCT/IB: Add DEVX support

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/base/ib_device.c [hunk 1]
```diff
@@ -310,11 +297,46 @@ ucs_status_t uct_ib_device_init(uct_ib_device_t *dev,
...
-            goto err_free_context;
-        }
+#if HAVE_DECL_IBV_EXP_DEVICE_DC_TRANSPORT && HAVE_STRUCT_IBV_EXP_DEVICE_ATTR_EXP_DEVICE_CAP_FLAGS
+    if (dev->dev_attr.exp_device_cap_flags & IBV_EXP_DEVICE_DC_TRANSPORT) {
+        dev->flags |= UCT_IB_DEVICE_FLAG_DC;
+    }
+#endif
+
+    *p_md = md;
+
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: redundant space

File: src/uct/ib/base/ib_md.h [hunk 2]
```diff
@@ -128,6 +128,29 @@ typedef struct uct_ib_rcache_region {
...
+/**
+ * IB device private initializer.
+ */
+typedef struct uct_ib_device_init_entry {
+    ucs_list_link_t             list;
+    ucs_status_t                (*init)(struct ibv_device *ibv_device,
+                                        uct_ib_md_t **p_md);
+} uct_ib_device_init_entry_t;
+
+#define UCT_IB_DEVICE_INIT(_init_fn, device_specific) \
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h]: s/device_specific/_priority/

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -1385,11 +1387,12 @@ ucs_status_t
 uct_ib_md_open(const char *md_name, const uct_md_config_t *uct_md_config, uct_md_h *md_p)
 {
     const uct_ib_md_config_t *md_config = ucs_derived_of(uct_md_config, uct_ib_md_config_t);
+    uct_ib_md_t *md = NULL;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: need to init status to UNSUPPORTED and check it (in addition to assertion on md)

File: src/uct/ib/base/ib_md.h [hunk 4]
```diff
@@ -128,6 +128,29 @@ typedef struct uct_ib_rcache_region {
 } uct_ib_rcache_region_t;
 
 
+/**
+ * IB device private initializer.
+ */
+typedef struct uct_ib_device_init_entry {
+    ucs_list_link_t             list;
+    ucs_status_t                (*init)(struct ibv_device *ibv_device,
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h]: s/init/md_open/
+ document what it does

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 5]
```diff
@@ -72,20 +70,50 @@ static ucs_status_t uct_ib_mlx5_check_dc(uct_ib_device_t *dev)
...
+        ucs_debug("mlx5dv_open_device(%s) failed: %m", ibv_get_device_name(ibv_device));
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    md = ucs_calloc(1, sizeof(*md), "ib_md");
+    if (md == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+    dev              = &md->dev;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: space line before

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 6]
```diff
@@ -72,20 +70,50 @@ static ucs_status_t uct_ib_mlx5_check_dc(uct_ib_device_t *dev)
...
+    ctx = mlx5dv_open_device(ibv_device, &dv_attr);
+    if (ctx == NULL) {
+        ucs_debug("mlx5dv_open_device(%s) failed: %m", ibv_get_device_name(ibv_device));
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    md = ucs_calloc(1, sizeof(*md), "ib_md");
+    if (md == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: goto err_close_device

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 7]
```diff
@@ -72,20 +70,50 @@ static ucs_status_t uct_ib_mlx5_check_dc(uct_ib_device_t *dev)
...
+    }
+    dev              = &md->dev;
+    dev->ibv_context = ctx;
+
+    IBV_EXP_DEVICE_ATTR_SET_COMP_MASK(&dev->dev_attr);
+    ret = ibv_query_device(dev->ibv_context, &dev->dev_attr);
+    if (ret != 0) {
+        ucs_error("ibv_query_device() returned %d: %m", ret);
+        status = UCS_ERR_IO_ERROR;
+        goto err;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: s/err/err_free/

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 8]
```diff
@@ -155,17 +187,22 @@ static ucs_status_t uct_ib_mlx5dv_device_init(uct_ib_device_t *dev)
             dev->ext_atomic_arg_sizes_be = arg_size;
         }
     }
-#endif
+
     if (has_dc) {
         status = uct_ib_mlx5_check_dc(dev);
     }
+    *p_md = md;
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: space before, instead of after

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8614/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.h [hunk 9]
```diff
@@ -129,24 +129,27 @@ typedef struct uct_ib_rcache_region {
...
     ucs_list_link_t             list;
-    ucs_status_t                (*init)(struct ibv_device *ibv_device,
-                                        uct_ib_md_t **p_md);
-} uct_ib_device_init_entry_t;
+    ucs_status_t                (*md_open)(struct ibv_device *ibv_device,
+                                           uct_ib_md_t **p_md);
+} uct_ib_md_open_entry_t;
 
-#define UCT_IB_DEVICE_INIT(_init_fn, device_specific) \
+#define UCT_IB_MD_OPEN(_open_fn, priority) \
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h]: _priority

File: src/uct/ib/base/ib_md.c [hunk 10]
```diff
@@ -1413,21 +1413,27 @@ uct_ib_md_open(const char *md_name, const uct_md_config_t *uct_md_config, uct_md
...
+        status = md_open_entry->md_open(ib_device, &md);
         if (status == UCS_OK) {
             break;
         } else if (status != UCS_ERR_UNSUPPORTED) {
             goto out_free_dev_list;
         }
     }
 
+    if (status != UCS_OK) {
+        ucs_debug("Failed to initialize IB device %s", md_name);
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: `ucs_assert(status == UCS_ERR_UNSUPPORTED)`
`ucs_debug("Unsupported IB device %s", md_name);`

File: src/uct/ib/base/ib_md.h [hunk 11]
```diff
@@ -129,24 +129,27 @@ typedef struct uct_ib_rcache_region {
...
-        extern ucs_list_link_t uct_ib_device_init_list; \
-        static uct_ib_device_init_entry_t init_entry = { \
-            .init = _init_fn, \
+        extern ucs_list_link_t uct_ib_md_open_list; \
+        static uct_ib_md_open_entry_t entry = { \
+            .md_open = _open_fn, \
         }; \
-        if (device_specific) { \
-            ucs_list_add_head(&uct_ib_device_init_list, &init_entry.list); \
+        if (priority == 1) { \
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.h]: if (priority)

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8617/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8619/ for details (Mellanox internal link).

**Pull request merged**